gson/src/main/java/com/google/gson/internal/bind/RecordTypeAdapterFactory.java

```diff
@@ -0,0 +1,480 @@
+            // do not support serialization of it. The contract for TypeAdapterFactory is to return null for
+            // unsupported types. Gson will then attempt other means of creating a TypeAdapter. If all fails, an
+            // exception will be created where Gson was invoked to serialize/deserialize a type.
+            return null;
+        }
+    }
+
+    /**
+     * Factory to construct our own {@link RecordField} instances. This requires some reflection into the internal
+     * of Gson, which has logic to handle the {@link JsonAdapter} annotation. We also need to reflect into Gson
+     * to fetch the {@link Excluder}, that handles the {@link com.google.gson.annotations.Expose},
+     * {@link com.google.gson.annotations.Since} and {@link com.google.gson.annotations.Until} annotations. This
+     * allows us to closely follow the Gson mechanisms wrt. serialization/deserialization.
+     */
+    private static final class RecordFieldFactory {
+
+
+        private final Gson gson;
+        private final Excluder excluder;
+        private final JsonAdapterAnnotationTypeAdapterFactory jsonAdapterFactory;
+        private final ConstructorConstructor constructorConstructor;
+
+        public RecordFieldFactory(Gson gson, Excluder excluder,
+                JsonAdapterAnnotationTypeAdapterFactory jsonAdapterFactory,
+                ConstructorConstructor constructorConstructor) {
+            this.gson = gson;
+            this.excluder = excluder;
+            this.jsonAdapterFactory = jsonAdapterFactory;
+            this.constructorConstructor = constructorConstructor;
+        }
+
+        private RecordField createRecordField(Field field, Object recordComponent) {
+            TypeToken<?> fieldType = TypeToken.get(RECORD_HELPER.getGenericType(recordComponent));
+            JsonAdapter jsonAdapter = RECORD_HELPER.getAnnotation(recordComponent, JsonAdapter.class);
+            TypeAdapter<?> typeAdapter;
+            if (jsonAdapter == null) {
+                typeAdapter = gson.getAdapter(fieldType);
+            }
+            else {
+                typeAdapter = jsonAdapterFactory.getTypeAdapter(
+                        constructorConstructor,
+                        gson,
+                        fieldType,
+                        jsonAdapter
+                );
+            }
+
+            // Determine the serialized and deserialized names. If there is an SerializedName annotation
+            // present, we need to respect it. This might also mean that multiple keys acts as aliases
+            // when reading Json.
+            SerializedName serializedNameAnnotation =
+                    RECORD_HELPER.getAnnotation(recordComponent, SerializedName.class);
+            String serializedName;
+            String[] deserializedNames;
+            if (serializedNameAnnotation == null) {
+                serializedName = RECORD_HELPER.getName(recordComponent);
```

Comment:
Maybe this should pass the private field to the `FieldNamingStrategy` to behave similarly to `ReflectiveTypeAdapterFactory`.